### PR TITLE
Feat/email login flow

### PR DESCRIPTION
**Pull request by alexanderleegs**
## Problem

This PR builds on https://github.com/isomerpages/isomercms-backend/pull/479, as part of our login flow release. To be reviewed in conjunction with PR #[996](https://github.com/isomerpages/isomercms-frontend/pull/996) on the isomercms-frontend repo. It introduces endpoints for the new login flow and modifies certain methods (site retrieval, commit messages) which were previously reliant on having a user's github id.
In addition, this PR also fixes a bug with our existing DB schema - we were previously associating Users to the Users table via the SiteMembers table, which has been corrected to associate the Users table with the Sites table. Aliases have also been added to the relations to allow for querying of related tables.

In summary, the following items have been implemented:

- New endpoints for verification and sending of email otp for login
- Updating of commit message to include user id of the person making the change
- Fixing of our existing database schema
- Modification of the existing users database table to allow for empty `githubId` field
- Separate flow for site retrieval for email and github users
- Updating of user by isomer id rather than github id

TBD before merging:

- Run db migration

### CODE REVIEW

**Comment by seaerchin** [comment]
@alexanderleegs i think the otp test seems to be failing - i'll wait for you to doublecheck this before reviewing!

**Comment by alexanderleegs** [comment]
> @alexanderleegs i think the otp test seems to be failing - i'll wait for you to doublecheck this before reviewing!

Thanks for catching, forgot to run tests before pushing! Have corrected the tests + made necessary bug fixes

File: src/services/utilServices/SitesService.js [hunk 1]
```diff
@@ -29,12 +29,33 @@ const ISOMER_ADMIN_REPOS = [
...
+  }) {
     this.githubService = gitHubService
     this.configYmlService = configYmlService
+    this.usersService = usersService
+    this.isomerAdminsService = isomerAdminsService
   }
 
   async getSites(sessionData) {
+    const isEmailUser = sessionData.getIsEmailUser()
+    let emailAccessibleSites = []
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:46]: nit: could we rename this to `retrievedSitesByEmail`? i got confused by this at first ._.

File: src/services/utilServices/SitesService.js [hunk 2]
```diff
@@ -29,12 +29,33 @@ const ISOMER_ADMIN_REPOS = [
...
+    if (!isAdminUser && isEmailUser) {
+      const user = await this.usersService.findSitesByUserId(userId)
+      if (user) {
+        const { site_members: siteMemberEntries } = user
+        emailAccessibleSites = siteMemberEntries.map((entry) => {
+          const repoData = entry.repo
+          return repoData.name
+        })
+      }
+    }
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:58]: this seems abit cluttered - wdyt about combining the nested `if` conditions to make this more clear? alternatively, we could shift this and the conditional check below into helper methods (maybe together with the retrieval?) which is fine too

File: src/services/utilServices/__tests__/AuthService.spec.js [hunk 3]
```diff
@@ -101,19 +116,76 @@ describe("Auth Service", () => {
...
+        service.verifyOtp({ email: mockEmail, otp: mockOtp })
+      ).rejects.toThrow(BadRequestError)
+      expect(mockUsersService.verifyOtp).toHaveBeenCalledWith(
+        mockEmail,
+        mockOtp
+      )
+    })
+  })
+
   describe("getUserInfo", () => {
```
Comment (seaerchin) [inline src/services/utilServices/__tests__/AuthService.spec.js:165]: is it possible for github and email to be both unset and if so, should we add a test case checking that this throws an error? 

should we also check that invalid github ids and emails return errors?

File: src/services/utilServices/__tests__/SitesService.spec.js [hunk 4]
```diff
@@ -24,18 +28,140 @@ describe("Resource Page Service", () => {
...
+
+      expect(mockIsomerAdminsService.getByUserId).toHaveBeenCalledWith(
+        mockIsomerUserId
+      )
+      expect(genericGitHubAxiosInstance.get).toHaveBeenCalledTimes(3)
+      process.env.ISOMERPAGES_REPO_PAGE_COUNT = currRepoCount
+      expect(process.env.ISOMERPAGES_REPO_PAGE_COUNT).toBe(currRepoCount)
+    })
+
+    it("Filters accessible sites for email user with no sitescorrectly", async () => {
```
Comment (seaerchin) [inline src/services/utilServices/__tests__/SitesService.spec.js:131]: nit: add spacing between sites and correctly

**Review by seaerchin — CHANGES_REQUESTED** [review]

File: src/fixtures/sessionData.ts [hunk 5]
```diff
@@ -19,5 +19,13 @@ export const mockSessionData = new SessionData({
...
+
+export const mockSessionDataEmailUser = new SessionData({
+  isomerUserId: mockIsomerUserId,
+  email: mockEmail,
+})
+
 mockSessionData.addGithubState(mockGithubState)
 mockSessionData.addSiteName(mockSiteName)
+mockSessionDataEmailUser.addGithubState(mockGithubState)
+mockSessionDataEmailUser.addSiteName(mockSiteName)
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/fixtures/sessionData.ts]: Yup, fixed in the refactor

File: src/routes/v2/authenticated/__tests__/Sites.spec.js [hunk 6]
```diff
@@ -59,28 +59,6 @@ describe("Sites Router", () => {
     })
   })
 
-  describe("checkHasAccess", () => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/routes/v2/authenticated/__tests__/Sites.spec.js:64]: Forgot to add new tests, thanks for catching! The equivalent tests have been introduced in https://github.com/isomerpages/isomercms-backend/pull/480/commits/a30c37d1f3e9427497fd4968d3509cb7c9477836

File: src/services/db/__tests__/GitHubService.spec.js [hunk 7]
```diff
@@ -596,6 +611,10 @@ describe("Github Service", () => {
       const secondSha = "second-sha"
       const gitTree = "git-tree"
       const message = "message"
+      const finalExpectedMessage = JSON.stringify({
+        message: message || `isomerCMS updated ${siteName} state`,
+        userId,
+      })
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/db/__tests__/GitHubService.spec.js]: You're right, this isn't needed - removed in https://github.com/isomerpages/isomercms-backend/pull/480/commits/0fa4fd60fde264238ddc7ca7c2dfbe651151005d

File: src/services/identity/UsersService.ts [hunk 8]
```diff
@@ -58,12 +58,41 @@ class UsersService {
...
+          required: true,
+          include: [
+            {
+              model: Repo,
+              required: true,
+              where: {
+                name: siteName,
+              },
+            },
+          ],
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/identity/UsersService.ts:74]: `Repo` is the database table containing `siteName`, which we need to use for comparison!

File: src/services/identity/UsersService.ts [hunk 9]
```diff
@@ -58,12 +58,41 @@ class UsersService {
...
 
+  async findSitesByUserId(isomerId: string) {
+    return this.repository.findOne({
+      where: { id: isomerId },
+      include: [
+        {
+          model: Site,
+          as: "site_members",
+          required: true,
+          include: [{ model: Repo, required: true }],
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/identity/UsersService.ts:90]: See above

File: src/services/utilServices/AuthService.js [hunk 10]
```diff
@@ -90,19 +92,49 @@ class AuthService {
...
-        headers: {
-          Authorization: `token ${accessToken}`,
-          "Content-Type": "application/json",
-        },
-      })
-      const userId = resp.data.login
+      if (sessionData.getIsEmailUser()) {
+        const email = sessionData.getEmail()
+        const { contactNumber } = await this.usersService.findByEmail(email)
+        return { userId: "", email, contactNumber }
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/utilServices/AuthService.js]: Makes sense, https://github.com/isomerpages/isomercms-backend/pull/480/commits/6c0561fc8643e077980d3ab382ef0f538901292f

File: src/services/utilServices/SitesService.js (hunk 1 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: Sure, https://github.com/isomerpages/isomercms-backend/pull/480/commits/0fdffecba8dfc13ac707441a01c702444935db99

File: src/services/utilServices/SitesService.js (hunk 2 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: Shifted into a helper method in https://github.com/isomerpages/isomercms-backend/pull/480/commits/88065efa8178db74d582fe792c7242f5780248f7

File: src/services/utilServices/__tests__/AuthService.spec.js (hunk 3 above)
Comment (alexanderleegs) [reply src/services/utilServices/__tests__/AuthService.spec.js:165]: It shouldn't be possible for both to be unset - this should be caught by an earlier middleware.

I think the invalid checks belongs in the called methods - otherwise we're just forcing the called method to throw an error

File: src/services/utilServices/__tests__/SitesService.spec.js (hunk 4 above)
Comment (alexanderleegs) [reply src/services/utilServices/__tests__/SitesService.spec.js]: Fixed in https://github.com/isomerpages/isomercms-backend/pull/480/commits/686087e5ee9a7cb687931ed23f0774783f633f9a

File: src/services/identity/UsersService.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/services/identity/UsersService.ts:74]: ah ok, gotcha! thanks for clarifying

File: src/services/utilServices/AuthService.js [hunk 11]
```diff
@@ -9,7 +9,9 @@ const { ForbiddenError } = require("@errors/ForbiddenError")
 const validateStatus = require("@utils/axios-utils")
```
Comment (seaerchin) [inline src/services/utilServices/AuthService.js:9]: i just realised this is also `AuthService` :monkas:

File: src/services/utilServices/SitesService.js [hunk 12]
```diff
@@ -29,12 +29,38 @@ const ISOMER_ADMIN_REPOS = [
...
+  async getEmailUserSites(userId) {
+    const user = await this.usersService.findSitesByUserId(userId)
+    if (user) {
+      const { site_members: siteMemberEntries } = user
+      return siteMemberEntries.map((entry) => {
+        const repoData = entry.repo
+        return repoData.name
+      })
+    }
+    return []
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:53]: we could flip the condition so this reads more smoothly

File: src/services/utilServices/SitesService.js [hunk 13]
```diff
@@ -74,6 +100,12 @@ class SitesService {
             isPrivate,
           }
         })
+        .filter(
+          (repoData) =>
+            isAdminUser ||
+            !isEmailUser ||
+            retrievedSitesByEmail.includes(repoData.repoName)
+        )
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:108]: could i ask why we're doing a `filter` operation here rather than just an early return above, when `isAdminUser` and `isEmailUser` is known? 

it seems like it'll still return an empty array and it would be better for readability too

File: src/services/utilServices/SitesService.js [hunk 14]
```diff
@@ -29,12 +29,38 @@ const ISOMER_ADMIN_REPOS = [
...
   }
 
   async getSites(sessionData) {
+    const isEmailUser = sessionData.isEmailUser()
+    let retrievedSitesByEmail = []
+    const { isomerUserId: userId } = sessionData
+    const isAdminUser = !!(await this.isomerAdminsService.getByUserId(userId))
+    if (!isAdminUser && isEmailUser)
+      retrievedSitesByEmail = await this.getEmailUserSites(userId)
+
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:63]: not sure if this is out of scope for this PR but wdyt about refactoring this to 2 methods called `getSitesForUser` and `getSitesForAdmin`? 

this method is p confusing to read through at the moment, especially with the nested `map/filter` operations ._.

File: src/services/utilServices/__tests__/AuthService.spec.js [hunk 15]
```diff
@@ -101,21 +116,77 @@ describe("Auth Service", () => {
...
+        service.verifyOtp({ email: mockEmail, otp: mockOtp })
+      ).resolves.toEqual(signedToken)
+      expect(mockUsersService.verifyOtp).toHaveBeenCalledWith(
+        mockEmail,
+        mockOtp
+      )
+      expect(mockUsersService.loginWithEmail).toHaveBeenCalledWith(mockEmail)
+    })
+
+    it("should throw an error if otp is incorrect", async () => {
```
Comment (seaerchin) [inline src/services/utilServices/__tests__/AuthService.spec.js:152]: hmm, is it possible to add a test case for expired otps? (lmk if this is diff to do)

**Review by seaerchin — COMMENTED** [review]
not sure if this is done but i saw that the changes to _integration_ tests were q minor and seemed to be compat fixes rather than true integration tests - do we already have integration tests for this and if not, should we do so, considering that this is a p important flow?

File: src/services/utilServices/SitesService.js (hunk 12 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: https://github.com/isomerpages/isomercms-backend/pull/480/commits/9607bdc51f19e84a9c33cb909573417825a46ad8

File: src/services/utilServices/SitesService.js (hunk 13 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: Hmm I think this part isn't very clear - the idea was for all sites retrieved to pass through if the user was an admin user, or using their own github token (since that only retrieves repos they have access to). I've refactored the code to make this clearer in https://github.com/isomerpages/isomercms-backend/pull/480/commits/86cdc315049dfd6dead583c715fefe02329822ce

File: src/services/utilServices/SitesService.js (hunk 14 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: Same as above, I think this part might have been confusing, so I refactored the code - isAdminUser is supposed to simply allow the result of the github api query to return all sites. Changed in https://github.com/isomerpages/isomercms-backend/pull/480/commits/86cdc315049dfd6dead583c715fefe02329822ce, lmk if it's still confusing

File: src/services/utilServices/__tests__/AuthService.spec.js (hunk 15 above)
Comment (alexanderleegs) [reply src/services/utilServices/__tests__/AuthService.spec.js:152]: I actually don't know if there's a special error message for expired otps - I've only been able to receive wrong otp error messages, so I don't have a test case for expired ones

File: src/services/utilServices/AuthService.js (hunk 11 above)
Comment (alexanderleegs) [reply src/services/utilServices/AuthService.js:9]: Do you think we should also rename this service? We could file an issue (and there's also an `AuthService` under identity)

File: src/routes/v2/authenticated/__tests__/Sites.spec.js [hunk 16]
```diff
@@ -61,28 +61,6 @@ describe("Sites Router", () => {
...
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/routes/v2/authenticated/__tests__/Sites.spec.js:64]: Yep

File: src/services/api/AxiosInstance.ts [hunk 17]
```diff
@@ -13,7 +13,11 @@ const requestFormatter = async (config: AxiosRequestConfig) => {
   const authMessage = config.headers.Authorization
 
   // If accessToken is missing, authMessage is `token `
-  if (!authMessage || authMessage === "token ") {
+  if (
+    !authMessage ||
+    authMessage === "token " ||
+    authMessage === "token undefined"
+  ) {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/api/AxiosInstance.ts:20]: I also don't really like this, but we either seem to be having a blank access token or undefined access token and this was the place which required the least code change to fix - i figured we could leave it as this is relatively temporary (until we move away from github login entirely)

**Comment by alexanderleegs** [comment]
> not sure if this is done but i saw that the changes to _integration_ tests were q minor and seemed to be compat fixes rather than true integration tests - do we already have integration tests for this and if not, should we do so, considering that this is a p important flow?

Hmm there isn't actually anything which really queries the database to require an integration test other than the new `getSites` - i've added an integration test for that 86cdc31

File: src/services/utilServices/SitesService.js (hunk 14 above)
Comment (seaerchin) [reply src/services/utilServices/SitesService.js]: double checking here - the reason why `!isEmailUser` is part of the condition (`isAdminuser || !isEmailUser`) is because we use the user's own github access token to retrieve the sites and github will resolve permissions for us on their end?

File: src/services/utilServices/SitesService.js (hunk 13 above)
Comment (seaerchin) [reply src/services/utilServices/SitesService.js]: i've made a comment below wrt this - i think it reads better now but the underlying assumption (that if they're github users, then they are using their own token, which will retrieve your owned sites and we can omit the `filter` on our end) should be documented here!

File: src/services/api/AxiosInstance.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (seaerchin) [reply src/services/api/AxiosInstance.ts:20]: yea, i think fixing this is beyond hte scope of this PR but just smth to keep in mind :(

File: src/services/utilServices/AuthService.js (hunk 11 above)
Comment (seaerchin) [reply src/services/utilServices/AuthService.js:9]: tbh, our naming is p confusing so i feel like we should; i think filing an issue might be kicking the can down the road so if it's convenient here, i'm more inclined to doing it here (TIA HAHAHAH)

File: src/services/utilServices/__tests__/AuthService.spec.js (hunk 15 above)
Comment (seaerchin) [reply src/services/utilServices/__tests__/AuthService.spec.js:152]: i looked through - there isn't; **all** verification is treated equally and we return the same error message even if the underlying error is different...

idt this is part of this PR (probably out of scope) but i think we should be aware of this and i'll file an issue accordingly @__@

File: src/integration/Sites.spec.ts [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+import { SitesService } from "@root/services/utilServices/SitesService"
+import { getUsersService } from "@services/identity"
+import { sequelize } from "@tests/database"
+
+const mockSite = "mockSite"
+const mockSiteId = "1"
+const mockBlockedSite = "adminOnly"
+const mockUpdatedAt = "now"
+const mockPermissions = { push: true }
+const mockPrivate = true
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:23]: i think we can omit this :P but if you wanna keep this, could we rename it to `mockIsPrivate`?

File: src/integration/Sites.spec.ts [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+      ;((genericGitHubAxiosInstance.get as unknown) as jest.Mock).mockImplementationOnce(
+        () => ({
+          data: [],
+        })
+      )
+      ;((genericGitHubAxiosInstance.get as unknown) as jest.Mock).mockImplementationOnce(
+        () => ({
+          data: [],
+        })
+      )
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:140]: actually i think this works b/c there's a default `axios` mock in `__mocks__` - we should be using that instead HAHA

File: src/integration/Sites.spec.ts [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:196]: should we consider using `mockAxios`?

File: src/integration/Sites.spec.ts [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+      })
+    })
+    afterEach(async () => {
+      // Clean up so that different tests using
+      // the same mock user don't interfere with each other
+      await IsomerAdmin.destroy({
+        where: { userId: mockIsomerUserId },
+      })
+    })
+    it("should return list of only sites available to user", async () => {
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:101]: nit: we should specify that this is an email user as github users have slightly diff behaviour

File: src/integration/Sites.spec.ts [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+subrouter.use(sitesSubrouter)
+const app = generateRouter(subrouter)
+
+describe("Sites Router", () => {
+  afterEach(() => {
+    jest.resetAllMocks()
+    mockAxios.reset()
+  })
+
+  describe("/", () => {
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:61]: could we have a few more test cases so that these invariants are clear

1. for github users, their access token only gives them access to their site
2. sites w/o `siteMembers` are seen by admins (i think this is so for `blockedSite`? if so, maybe just rename it to make it clearer)
3. public sites can be seen by everyone (<-- not q sure what should be the behaviour here, correct me if i'm wrong)

File: src/services/utilServices/SitesService.js [hunk 23]
```diff
@@ -47,7 +68,7 @@ class SitesService {
       })
     )
 
-    const sites = await Bluebird.map(paramsArr, async (params) => {
+    const allSites = await Bluebird.map(paramsArr, async (params) => {
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:71]: nit: `allSites` reads slightly weird but as i have no good name, i'm ok w/ just leaving this as is ._.

File: src/services/utilServices/AuthService.js (hunk 11 above)
Comment (alexanderleegs) [reply src/services/utilServices/AuthService.js:9]: It would probably involve frontend changes as well (endpoints are currently located in `/auth/...`) - hence why i think this would be better to handle in a separate PR

File: src/integration/Sites.spec.ts (hunk 18 above)
Comment (alexanderleegs) [reply src/integration/Sites.spec.ts:23]: Hmm i was thinking this is useful because we can check that regular users only have access to items in the site members database that include them, while admin users see all by default - I think it might be better to leave in, but I'm not sure if `mockIsPrivate` conveys that idea

File: src/integration/Sites.spec.ts (hunk 19 above)
Comment (alexanderleegs) [reply src/integration/Sites.spec.ts:140]: I tried that earlier actually, but couldn't get it to work - I think it might have something to do with the `AxiosInstance` defined elsewhere which uses both `create` and interceptors. This was the way I got it to work with `genericGithubAxiosInstance`, but do let me know if it can be done with `mockAxios` instead!

File: src/integration/Sites.spec.ts (hunk 22 above)
Comment (alexanderleegs) [reply src/integration/Sites.spec.ts:61]: 1. I left this one out of the integration tests because it's covered by the unit tests - the Github API call is the one which restricts the sites accessible to the user, so I didn't think there was much benefit in including it here - we'd just be mocking the Github results and seeing that we receive the same mocked results.
2. Yup that was the purpose of `blockedSite` - I've renamed it to `mockAdminSite` if that's clearer? in 1a88af5, lmk if there's a better name for it
3. Hmm I think there's no such thing as public sites really - sites are "public" in that anyone can view them, but access to sites are still restricted based on whether someone has _write_ permissions to the site, which is what the getSites method is returning

File: src/services/utilServices/SitesService.js (hunk 13 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js]: Makes sense, added comment in 85aa065

File: src/integration/Sites.spec.ts (hunk 21 above)
Comment (alexanderleegs) [reply src/integration/Sites.spec.ts]: Fixed in https://github.com/isomerpages/isomercms-backend/pull/480/commits/1a88af565617b2f51cb3932b664c811902078c4c!

File: CHANGELOG.md [hunk 24]
```diff
@@ -4,8 +4,26 @@ All notable changes to this project will be documented in this file. Dates are d
 
 Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).
 
+#### [v0.10.1](https://github.com/isomerpages/isomercms-backend/compare/v0.10.0...v0.10.1)
```
Comment (seaerchin) [inline CHANGELOG.md:7]: assuming this is beacuse of merge?

File: src/integration/Sites.spec.ts (hunk 22 above)
Comment (seaerchin) [reply src/integration/Sites.spec.ts:61]: yeap sure, 1/2 seems reasonable! so just double checking for 3 - when we call `getSites`, this is restricted based on _write_ access and not _read_ access? this is because all of our sites are public (on github) and when we say write access, in this context what it means is **users who are members of that github organization that own the site**? think the semantics of this needs to be nailed down HAHA

File: src/integration/Sites.spec.ts [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+          },
+          {
+            lastUpdated: mockUpdatedAt,
+            repoName: mockAdminSite,
+            isPrivate: mockPrivate,
+            permissions: mockPermissions,
+          },
+        ],
+      }
+      ;((genericGitHubAxiosInstance.get as unknown) as jest.Mock).mockImplementation(
```
Comment (seaerchin) [inline src/integration/Sites.spec.ts:169]: this actually impacts readability pretty significantly .-. do we have a better way around this? given that this is a test file, i'm not too concerned by it but if there's a quick way to rewrite this, i think we should

File: src/services/utilServices/SitesService.js [hunk 26]
```diff
@@ -81,7 +102,16 @@ class SitesService {
         )
     })
 
-    return _.flatten(sites)
+    const flattenedAllSites = _.flatten(allSites)
+    // Github users are using their own access token, which already filters sites to only those they have write access to
+    if (isAdminUser || !isEmailUser) return flattenedAllSites
```
Comment (seaerchin) [inline src/services/utilServices/SitesService.js:107]: sanity check here - should this be `!isAdminUser || !isEmailUser)` or simply `!isEmailUser`? from the comment, this seems like it should be a non-admin user using their own github access token (which follows the same flow as an admin user)

**Review by seaerchin — COMMENTED** [review]
some comments, overall looks good!

File: src/integration/Sites.spec.ts (hunk 19 above)
Comment (seaerchin) [reply src/integration/Sites.spec.ts]: resolved this - i left a comment in the code for this but the tl;dr is that `SitesService` is hitting the `genericGithubAxiosInstance`, which is a named export. hence, this requires a mock of the named export after which everything will work fine

File: src/services/utilServices/AuthService.js (hunk 11 above)
Comment (seaerchin) [reply src/services/utilServices/AuthService.js:9]: sure, this sounds reasonable - could we file a tracking issue?

File: src/integration/Sites.spec.ts (hunk 22 above)
Comment (alexanderleegs) [reply src/integration/Sites.spec.ts:71]: Yup! To clarify - the github repos are all public, and anyone can access them via github directly. On our CMS, only sites that a user is allowed to write to are viewable - for github login users, that's via checking push permissions (`repoData.permissions.push === true`), and for email login, via checking the `site_members` table

File: src/services/utilServices/SitesService.js (hunk 26 above)
Comment (alexanderleegs) [reply src/services/utilServices/SitesService.js:108]: Hmm the check is correct - to clarify, the flows are:

1. Email user - uses our centralised access token, so flattenedAllSites returns all sites, so we need to further restrict based on our database
2. Email user (admin) - uses our centralised access token, so flattenedAllSites returns all sites.
3. Github user - uses their own access token, so flattenedAllSites returns only those they have write access to
4. Github user (admin) - uses their own access token, so flattenedAllSites returns only those they have write access to (which is all sites)

I've added an extra comment about admin users in f23cbd2!

File: src/integration/Sites.spec.ts (hunk 22 above)
Comment (seaerchin) [reply src/integration/Sites.spec.ts:71]: ok, thanks for clarifying!

File: src/services/utilServices/SitesService.js (hunk 26 above)
Comment (seaerchin) [reply src/services/utilServices/SitesService.js:108]: gotcha i think i misunderstood the conditional as github login is caught by the `!isEmailUser` portion

**Review by seaerchin — APPROVED** [review]
lgtm

File: src/services/utilServices/AuthService.js (hunk 11 above)
Comment (alexanderleegs) [reply src/services/utilServices/AuthService.js:9]: Yup, issue filed https://github.com/isomerpages/isomercms-backend/issues/498
